consensus/src/units/validator.rs

```diff
@@ -0,0 +1,337 @@
+                pu
+            ),
+        }
+    }
+}
+
+impl<H: Hasher, D: Data, S: Signature> From<SignatureError<FullUnit<H, D>, S>>
+    for ValidationError<H, D, S>
+{
+    fn from(se: SignatureError<FullUnit<H, D>, S>) -> Self {
+        ValidationError::WrongSignature(se.unchecked)
+    }
+}
+
+pub struct Validator<'a, KB: KeyBox> {
+    session_id: SessionId,
+    keychain: &'a KB,
+    max_round: Round,
+    threshold: NodeCount,
+}
+
+type Result<'a, H, D, KB> =
+    StdResult<SignedUnit<'a, H, D, KB>, ValidationError<H, D, <KB as KeyBox>::Signature>>;
+
+impl<'a, KB: KeyBox> Validator<'a, KB> {
+    pub fn new(
+        session_id: SessionId,
+        keychain: &'a KB,
+        max_round: Round,
+        threshold: NodeCount,
+    ) -> Self {
+        Validator {
+            session_id,
+            keychain,
+            max_round,
+            threshold,
+        }
+    }
+
+    pub fn validate_unit<H: Hasher, D: Data>(
+        &self,
+        uu: UncheckedSignedUnit<H, D, KB::Signature>,
+    ) -> Result<'a, H, D, KB> {
+        let su = uu.check(self.keychain)?;
+        let full_unit = su.as_signable();
+        if full_unit.session_id() != self.session_id {
+            // NOTE: this implies malicious behavior as the unit's session_id
+            // is incompatible with session_id of the message it arrived in.
+            return Err(ValidationError::WrongSession(full_unit.clone()));
+        }
+        if full_unit.round() > self.max_round {
+            return Err(ValidationError::RoundTooHigh(full_unit.clone()));
+        }
+        self.validate_unit_parents(su)
+    }
+
+    fn validate_unit_parents<H: Hasher, D: Data>(
+        &self,
+        su: SignedUnit<'a, H, D, KB>,
+    ) -> Result<'a, H, D, KB> {
+        // NOTE: at this point we cannot validate correctness of the control hash, in principle it could be
+        // just a random hash, but we still would not be able to deduce that by looking at the unit only.
+        let pre_unit = su.as_signable().as_pre_unit();
+        if pre_unit.n_members() != self.keychain.node_count() {
+            return Err(ValidationError::WrongNumberOfMembers(pre_unit.clone()));
+        }
+        let round = pre_unit.round();
+        let n_parents = pre_unit.n_parents();
+        if round == 0 && n_parents > NodeCount(0) {
+            return Err(ValidationError::RoundZeroWithParents(pre_unit.clone()));
+        }
+        let threshold = self.threshold;
+        if round > 0 && n_parents < threshold {
+            return Err(ValidationError::NotEnoughParents(pre_unit.clone()));
+        }
+        let control_hash = &pre_unit.control_hash();
+        if round > 0 && !control_hash.parents_mask[pre_unit.creator()] {
+            return Err(ValidationError::NotDescendantOfPreviousUnit(
+                pre_unit.clone(),
+            ));
+        }
+        Ok(su)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I guess there are no checks for `WrongSignature`, `RoundZeroWithParents` and `NotDescendantOfPreviousUnit`

Reply:
Unfortunately. We cannot create wrong signatures with the current version of mocks (and I didn't want to modify them significantly before they get factored out), and we cannot use `Creator` to create the other scenarios. Since there were no tests at all before this refactor I hope I can sneak by with subpar coverage. <_<